### PR TITLE
Add option to prevent LDAP from deactivating everything on empty search

### LINKED ISSUE #7949: Gitea deactivates LDAP users
- Gitea version (or commit ref): 1.9.1 
- Git version: 2.22.0
- Operating system: CentOS 7.6
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [X] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [X] Not relevant
- Log gist:

## Description
While using Gitea with FreeIPA as authentication source, Gitea repeatedly deactivates valid user account every time auth source is synchronised. This is rather annoying to the extent that it makes Gitea unusable in a FreeIPA environment.
There are already several issues opened in the past for this very same issue, but they either went stale or have been closed pointing to a hack that does not quite address the root cause of the issue. This is why I am creating this meta-issue referring to previous closed issues. If this issue also gets closed without a solution in sight, I will have to abandon using Gitea and move our stuff over to something else. :-( Making a desperate attempt to keep using Gitea because it is what I want.

Past issues: 
1. https://github.com/go-gitea/gitea/issues/4067  State = Closed ; Reason = a hack to NOT sync LDAP users was supplied.
2. https://github.com/go-gitea/gitea/issues/4433  State = Stale & Closed
3. https://github.com/go-gitea/gitea/issues/4404  State = Stale & Closed
4. https://github.com/go-gitea/gitea/issues/4402  State = Stale & Closed
5. https://github.com/go-gitea/gitea/issues/3815   State = Open

Issue # 3815 refers to LDAP call failing leading to users being de-activated. I cannot verify whether this is happening in my case - the logs don't quite show any LDAP activity in the first place.. I tried running in dev instead of prod as well without luck.

Even if LDAP call fails, de-activating users should not be the default activity! Even in good conditions LDAP server may go offline temporarily, which will lead to disruption in Gitea unnecessarily. This feature needs to be modified such that activation/de-activation of users happen only on a successful LDAP call. 

Alternatively, a new config option may be added for e.g local_activation = true which skips the activation deactivation field of the SQL query if the config var is set, and allow the Gitea administrator to activate or de-activate all users(local+LDAP) in Gitea admin panel.

Why am I not okay with the hack? 
The hack is to supply `UPDATE_EXISTING = false` for LDAP sync. I have users who changed their passwords in FreeIPA, and Gitea will not take the new password. As an admin, I do not have the ability to sync this owner's LDAP data alone. I end up re-enabling LDAP sync and resort to manually activate all users after that 👎 only for the same to happen the next time a user changes their password.

Happy to provide any information from my installation. Desperately looking for help. 
...

### PR DESCRIPTION
**Pull request by zeripath**
It appears that some LDAPs will simply return an empty search results if there is an error during searching rather than report the error. In those circumstances Gitea would completely deactivate all LDAP users.

This is clearly not ideal - add a configuration option to LDAP configuration which will mean that by default an empty search results will not deactivate any users and instead an error will be logged. An authentication option can be turned on for those situations where this behaviour would be intended.

Fix #7949

### CODE REVIEW

File: options/locale/locale_en-US.ini [hunk 1]
```diff
@@ -1851,6 +1851,7 @@ auths.attribute_surname = Surname Attribute
 auths.attribute_mail = Email Attribute
 auths.attribute_ssh_public_key = Public SSH Key Attribute
 auths.attributes_in_bind = Fetch Attributes in Bind DN Context
+auths.allow_deactivate_all = Allow an empty search update result to deactivate all users
```
Comment (guillep2k) [inline options/locale/locale_en-US.ini]:
```suggestion
auths.allow_deactivate_all = Allow an empty search result to deactivate all users
```

**Review by guillep2k — COMMENTED** [review]
This should require a migration step, I think. What should the migration value be? I vote for `AllowDeactivateAll=false`.

**Comment by zeripath** [comment]
It's actually not in the db like that. It's JSON so it will default to false already.

**Review by guillep2k — APPROVED** [review]

**Review by 6543 — APPROVED** [review]

File: modules/auth/ldap/ldap.go [hunk 2]
```diff
@@ -47,6 +47,7 @@ type Source struct {
 	Filter                string // Query filter to validate entry
 	AdminFilter           string // Query filter to check if user is admin
 	Enabled               bool   // if this source is disabled
+	AllowDeactivateAll    bool   // Allow an empty search response to deactivate all users from this source
```
Comment (lunny) [inline modules/auth/ldap/ldap.go:50]: ~You should add migrations~

**Review by lunny — CHANGES_REQUESTED** [review]

**Review by lunny — APPROVED** [review]

**Comment by dioss-Machiel** [comment]
Is it at all possible to get this fix into a 1.10 release?

**Comment by zeripath** [comment]
We could probably backport it to 1.11?

I think 1.10 is now security fixes only. `git cherry-pick c5014a7` might work to put it on 1.10 but it will likely conflict.

**Comment by guillep2k** [comment]
> 
> 
> We could probably backport it to 1.11?
> 
> I think 1.10 is now security fixes only. `git cherry-pick c5014a7` might work to put it on 1.10 but it will likely conflict.

Sounds like we're close to 1.11, but if anybody wants port this to 1.10.4 and fix the conflicts or do it by hand, I have no problem reviewing the PR from zero.

**Comment by lunny** [comment]
@zeripath If we marked this as feature, I don't think we should send backport to v1.11. Or we may should label this as bug.

**Comment by zeripath** [comment]
@dioss-Machiel Looks like it applied cleanly to release/v1.10?

**Comment by dioss-Machiel** [comment]
I applied it manually, seems to be fine.

**Comment by zeripath** [comment]
I've marked this as a bug. I'm not sure where the bug is because LDAP shouldn't be returning an empty result without an error.

**Pull request merged**